security/advancedtls/provider.go

```diff
@@ -0,0 +1,151 @@
+// and root certificates based on the input PEM files.
+type PEMFileProvider struct {
+	*certprovider.Distributor
+	cancel context.CancelFunc
+}
+
+// NewPEMFileProvider uses PEMFileProviderOptions to construct a PEMFileProvider.
+func NewPEMFileProvider(o *PEMFileProviderOptions) (*PEMFileProvider, error) {
+	var identityUpdate = false
+	var rootUpdate = false
+	if o.CertFile != "" && o.KeyFile != "" {
+		identityUpdate = true
+	} else if o.CertFile != "" || o.KeyFile != "" {
+		logger.Warning("users must specify both KeyFile and CertFile to update identity credentials")
+	}
+	if o.TrustFile != "" {
+		rootUpdate = true
+	}
+	if o.IdentityInterval == 0 {
+		// If IdentityInterval is not set by users explicitly, we will set it to default interval.
+		o.IdentityInterval = defaultInterval
+	}
+	if o.RootInterval == 0 {
+		// If RootInterval is not set by users explicitly, we will set it to default interval.
+		o.RootInterval = defaultInterval
+	}
+	identityTicker := time.NewTicker(o.IdentityInterval)
+	rootTicker := time.NewTicker(o.RootInterval)
+	ctx, cancel := context.WithCancel(context.Background())
+	provider := &PEMFileProvider{Distributor: certprovider.NewDistributor()}
+	// A goroutine to pull file changes.
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				identityTicker.Stop()
+				rootTicker.Stop()
+				return
+			case <-identityTicker.C:
+				if !identityUpdate {
+					continue
+				}
+				// Read identity certs from PEM files.
+				identityCert, err := tls.LoadX509KeyPair(o.CertFile, o.KeyFile)
+				if err != nil {
+					// If the reading produces an error, we will skip the update for this round and log the error.
+					// Note that LoadX509KeyPair will return error if file is empty,
+					// so there is no separate check for empty file contents.
+					logger.Warning("tls.LoadX509KeyPair(%v, %v) failed: %v", o.CertFile, o.KeyFile, err)
+					continue
+				}
+				provider.Set(&certprovider.KeyMaterial{Certs: []tls.Certificate{identityCert}}, nil)
+			case <-rootTicker.C:
+				if !rootUpdate {
+					continue
+				}
+				// Read root certs from PEM files.
+				trustData, err := ioutil.ReadFile(o.TrustFile)
+				if err != nil {
+					// If the reading produces an error, we will skip the update for this round and log the error.
+					logger.Warning("ioutil.ReadFile(%v) failed: %v", o.TrustFile, err)
+					continue
+				}
+				if len(trustData) == 0 {
+					// If the current file is empty, skip the update for this round.
+					continue
+				}
+				trustPool := x509.NewCertPool()
+				trustPool.AppendCertsFromPEM(trustData)
+				provider.Set(&certprovider.KeyMaterial{Roots: trustPool}, nil)
+			default:
+			}
+		}
+	}(ctx)
+	provider.cancel = cancel
+	return provider, nil
+}
+
+// Close cleans up resources allocated by the PEMFileProvider.
+func (p *PEMFileProvider) Close() {
+	p.cancel()
+	p.Stop()
+}
+
+func getFileSize(filepath string) (int64, error) {
```

Comment:
Where is this used? I don't see any usages.

Reply:
Oh yeah, this was used previously, and I'll get rid of this.